core/commonMain/src/kotlinx/serialization/encoding/ReorderingCompositeEncoder.kt

```diff
@@ -0,0 +1,240 @@
+ */
+
+package kotlinx.serialization.encoding
+
+import kotlinx.serialization.*
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.modules.*
+
+
+/**
+ * Encodes composite elements in a specific order managed by [mapElementIndex].
+ *
+ * This encoder will replicate the behavior of a standard encoding, but calling the `encode*Element` methods in
+ * the order defined by [mapElementIndex]. It first buffers each `encode*Element` calls in an array following
+ * the given indexes using [mapElementIndex], then when [endStructure] is called, it encodes the buffered calls
+ * in the expected order by replaying the previous calls on a concrete [CompositeEncoder] provided
+ * by [beginStructure].
+ *
+ * @param structureDescriptor descriptor of the structure being encoded
+ * @param beginStructure provides the [CompositeEncoder] to be used to encode the given descriptor's elements in the expected order.
+ * @param mapElementIndex maps the element index to the new reordered index (zero-based). If this mapper provides the same index for multiple elements, only the last one will be encoded as the previous ones will be overridden.
+ */
+@ExperimentalSerializationApi
+internal class ReorderingCompositeEncoder(
+    structureDescriptor: SerialDescriptor,
+    private val beginStructure: (SerialDescriptor) -> CompositeEncoder,
+    private val mapElementIndex: (SerialDescriptor, Int) -> Int,
+) : CompositeEncoder {
+    // Each time we encode a field, if the next expected schema field index is not the good one, it is buffered until it's the time to encode it
+    private var bufferedCalls = Array<BufferedCall?>(structureDescriptor.elementsCount) { null }
+    private var callShouldEncodeElementDefault = false
+
+    override val serializersModule: SerializersModule
+        // No need to return a serializers module as it's not used during buffering
+        get() = EmptySerializersModule()
+
+    private data class BufferedCall(
+        val originalElementIndex: Int,
+        val encoder: CompositeEncoder.() -> Unit,
+    )
+
+    private inline fun bufferEncoding(
+        descriptor: SerialDescriptor,
+        index: Int,
+        crossinline encoder: CompositeEncoder.() -> Unit
+    ) {
+        if (callShouldEncodeElementDefault) {
+            bufferedCalls[mapElementIndex(descriptor, index)] = BufferedCall(index) {
+                if (shouldEncodeElementDefault(descriptor, index)) {
+                    encoder()
+                }
+            }
+        } else {
+            bufferedCalls[mapElementIndex(descriptor, index)] = BufferedCall(index) {
+                encoder()
+            }
+        }
+        callShouldEncodeElementDefault = false
+    }
+
+    override fun endStructure(descriptor: SerialDescriptor) {
+        encodeBufferedFields(descriptor)
+    }
+
+    private fun encodeBufferedFields(descriptor: SerialDescriptor) {
+        val classEncoder = beginStructure(descriptor)
```

Comment:
This can be called before creating this compositeEncoder in the call to `beginStructure` that returns this reordered one.

Reply:
Will be replaced by directly the encoder (no supplier)